### DISCUSSION TITLE
Support for other databases

### DISCUSSION THREAD

**Discussion by jogibear9988** [Giorgi/EFCore.Visualizer#23] [Ideas]
MySQL:
https://dev.mysql.com/doc/refman/8.0/en/explain-output.html
(Maria Db https://mariadb.com/kb/en/explain/)

SqLite:
https://www.sqlite.org/eqp.html

Oracle:
https://blogs.oracle.com/optimizer/post/how-to-generate-a-useful-sql-execution-plan

**Comment by Giorgi** [comment]
Thanks for the suggestion @jogibear9988. The idea behind this extension was to provide a graphical plan and not a text-based plan but if you or others send a PR with textual plan I will gladly accept it.

**Reply by jogibear9988** [reply]
I thought you create it from a text document. Didn't look exactly how it worked.

**Reply by Giorgi** [reply]
I do but I use other open-source projects to convert the text plan to visual representation. I wasn't able to find anything like that for the database engines you suggested.

**Comment by jogibear9988** [comment]
Short idea:

<img width="824" alt="image" src="https://github.com/Giorgi/EFCore.Visualizer/assets/364896/415900b7-aa5c-4bc3-adc1-7588b5f16b91">

		using System;
		using System.Collections.Generic;
		using System.Linq;
		using System.Text;

		public class SQLiteQueryPlanToGraphviz
		{
			public static string ConvertToGraphviz(string queryPlan)
			{
				StringBuilder dotGraph = new StringBuilder();
				dotGraph.AppendLine("digraph G {");
				dotGraph.AppendLine("  node [shape=box];"); // Set the node shape to box, typical for Graphviz diagrams

				var lines = queryPlan.Split(new[] { '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);
				var parentIdStack = new Stack<string>();
				int nodeIdCounter = 0;

				foreach (var line in lines)
				{
					var indentLevel = line.TakeWhile(c => c == ' ').Count() / 3; // Assuming 3 spaces per indent level for simplicity
					var trimmedLine = line.Trim();
					var nodeId = $"node{nodeIdCounter++}";

					// Create a label for the node that includes the query plan operation
					dotGraph.AppendLine($"  {nodeId} [label=\"{trimmedLine.Replace("\"", "\\\"")}\"];");

					// If the stack has an item at the current indent level, pop it (and any deeper items)
					while (parentIdStack.Count > indentLevel)
					{
						parentIdStack.Pop();
					}

					// If there's a parent node, create an edge from it to the current node
					if (parentIdStack.TryPeek(out string parent))
					{
						dotGraph.AppendLine($"  {parent} -> {nodeId};");
					}

					// Push the current node onto the stack
					parentIdStack.Push(nodeId);
				}

				dotGraph.AppendLine("}");

				return dotGraph.ToString();
			}
		}

		class Program
		{
			static void Main()
			{
				string queryPlan = @"`--COMPOUND QUERY
		   |--LEFT-MOST SUBQUERY
		   |  `--SCAN t1 USING COVERING INDEX i1
		   `--UNION USING TEMP B-TREE
			  `--SCAN t2 USING COVERING INDEX i4";
				string graphvizDot = SQLiteQueryPlanToGraphviz.ConvertToGraphviz(queryPlan);
				Console.WriteLine(graphvizDot);
			}
		}


rendered with graphviz:

<img width="820" alt="image" src="https://github.com/Giorgi/EFCore.Visualizer/assets/364896/943f35cd-2695-4186-9ccf-55986686e3eb">

**Reply by jogibear9988** [reply]
It's only a idea... still will need more work :-)

**Reply by jogibear9988** [reply]
<img width="818" alt="image" src="https://github.com/Giorgi/EFCore.Visualizer/assets/364896/2fa538f5-32b1-4697-81ca-a07799920120">

**Reply by Giorgi** [reply]
That's a good idea 👍

**Reply by Giorgi** [reply]
The image looks wrong though, `SCAN t1 USING COVERING INDEX i1` should be under `LEFT-MOST SUBQUERY`

**Comment by AzubuikeKevin** [comment]
Looking forward to MySQL been supported.

**Comment by Giorgi** [comment]
Oracle support is available in version 1.0.9

**Comment by Giorgi** [comment]
MySQL and SQLite support are planned for the next month or two.

**Comment by jogibear9988** [comment]
How will you do the support?

**Comment by Giorgi** [comment]
Version 1.1.10 adds support for SQLite:

![image](https://github.com/user-attachments/assets/3981c3cc-4307-4364-b79c-2cc1b50847bc)

**Comment by Giorgi** [comment]
Just published a new version with MySQL support. Check it out!

**Reply by jonataspc** [reply]
I'm getting this error when using MySQL:

<img width="395" height="152" alt="image" src="https://github.com/user-attachments/assets/af42535f-93e7-44a3-b760-8a3df9a493c5" />

MySQL Server version used: 8.0.43 

Looks like it only supports versions >= 8.4 since that variable was introduced in 8.4, can you confirm?

**Reply by Giorgi** [reply]
@jonataspc It was introduced in MySQL 8.3
